Group samples in DrILL interface

gui-co opened the pull request:
**Description of work.**

It is now possible to group samples in DrILL and set a 'master sample' within a group.

During processing, the master row parameter values will be used for all samples in that group. For a non master sample, a parameter value can still be overridden locally.

***example:***
Sample | Group | parameter 1 | parameter 2
-----------|----------|------------------|-----------------
1(master) |  g1 | v1 | v2
2              |  g1 |
3              |  g1|  | v2'

Here, all samples are in the same group and sample 1 is the master sample:
* during the processing of sample 2: `parameter 1 = v1` and `parameter 2 = v2`
* during the processing of sample 3: `parameter 1 = v1` and `parameter 2 = v2'`

**To test:**

With data from `ExternalData/Testing/Data/DocTest/ILL/D11/`.

* save the following json as `test.mrd`:
<details><summary>test.mrd</summary><p>

```json
{
    "Instrument": "D11",
    "AcquisitionMode": "SANS",
    "VisualSettings": {
        "FoldedColumns": [
            "SampleTransmissionRuns",
            "AbsorberRuns",
            "BeamRuns",
            "FluxRuns",
            "ContainerRuns",
            "ContainerTransmissionRuns",
            "TransmissionAbsorberRuns",
            "TransmissionBeamRuns",
            "MaskFiles",
            "ReferenceFiles",
            "OutputWorkspace"
        ],
        "HiddenColumns": [],
        "ColumnsOrder": [
            "SampleRuns",
            "SampleTransmissionRuns",
            "AbsorberRuns",
            "BeamRuns",
            "FluxRuns",
            "ContainerRuns",
            "ContainerTransmissionRuns",
            "TransmissionAbsorberRuns",
            "TransmissionBeamRuns",
            "MaskFiles",
            "ReferenceFiles",
            "OutputWorkspace",
            "SampleThickness",
            "CustomOptions"
        ]
    },
    "GlobalSettings": {
        "ThetaDependent": true,
        "SensitivityMaps": "",
        "DefaultMaskFile": "",
        "NormaliseBy": "Timer",
        "SampleThickness": 0.1,
        "BeamRadius": [
            0.1
        ],
        "TransmissionBeamRadius": 0.1,
        "WaterCrossSection": 1.0,
        "OutputType": "I(Q)",
        "CalculateResolution": "None",
        "DefaultQBinning": "PixelSizeBased",
        "BinningFactor": 1.0,
        "OutputBinning": [],
        "NPixelDivision": 1,
        "NumberOfWedges": 0,
        "WedgeAngle": 30.0,
        "WedgeOffset": 0.0,
        "AsymmetricWedges": false,
        "MaxQxy": [
            -1.0
        ],
        "DeltaQ": [
            -1.0
        ],
        "IQxQyLogBinning": false,
        "OutputPanels": false
    },
    "Samples": [
        {
            "SampleRuns": "2889"
        },
        {
            "SampleRuns": "2889",
            "SampleThickness": "0.2"
        },
        {
            "SampleRuns": "2889"
        }
    ],
    "SamplesGroups": {
        "A": [
            1,
            2
        ]
    },
    "MasterSamples": {
        "A": 1
    }
}
```
</p></details>

* open DrILL (Interfaces -> ILL -> Drill)
* load `test.mrd` (File -> Load rundex)
* observe the table: first row is not in a group, second and third rows are in a group and second row is the master of that group
* process all rows (Process -> Process all rows)
* check `SampleThickness` value in each output workspace history. `Sample_1 `: 0.1, `Sample_2`: 0.2, `Sample_3`: 0.2


Part of #29570 

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gvardany commented:
It worked with your example. However there are few problems:
- I created 6 rows, grouped 2 & 3, and 5&6, then tried to set 4 as master, and it created another group C, but row 1 I couldn't set as master since it doesn't belong to any group. We should make sure that a row that doesn't belong to a group cannot be set as master, there should also be a warning when trying to do so.
- Secondly, there seem to be a history accumulating, I played for some time, then erased the table and started over, and when saving rundex I saw there were groups that were no longer in the table. When do you clean the groups? I think a group should be deleted when there is no more row present in it. This might actually be related to the first problem.
- What happens when we reach letter Z and beyond?
- unrelated, the error dialog (in attachment) is not nice, can you change this to `Unable to process the rows ... Please check the logs.`
![Screenshot 2020-10-27 at 16 07 30](https://user-images.githubusercontent.com/16865759/97322143-c924ca00-186f-11eb-8f8a-ac3b7b848cdc.png)
- again unrelated, when clicking on process selected, it should run process all if there is nothing selected
- can you add process selected group(s) button in between with skip-forward icon, this should run only if at least one row is selected that belongs to a group
- can you make the rows drag and drop sortable just as columns? when creating a group with scattered rows one might want to visually bring them together too
- make a tooltip on the row heading (leftmost column), if it is in a group, tooltip should say `This row belongs to the sample group B`. if a row is a master (bold), the tooltip should say `This is the master row of the group D`.
- finally, this needs a detailed documentation; it's a nice feature, but can be overwhelming for the lambda users.

gui-co commented:
Here are the modifications:
* the history of groups is cleared with the table. When a sample is deleted, it is also removed from any group
* the group label are incremented from A to Z, then from AA to ZZ and so on. If a group has been deleted, its label becomes free and will be used for any new group
* the dialog message has been changed. If a problematic row belongs to a group, its label will be used in the message
* when clicking on `process selected row(s)`, if there is no selection, all row(s) are processed
* a new button allows the user to process the group(s) of the selected row(s)
* when grouping rows or setting a row as master row, the label is changed and a tooltip is added to explain the status of the row.
* the documentation has been added
* the export and import as rundex file code has been moved to a new class: `DrillRundexIO`

After discussion, it appears that the rows should not drag-and-drop sortable. One can always cut and paste rows and add a row at the middle of the table.

gvardany commented on docs/source/interfaces/DrILL.rst:
```diff
@@ -131,6 +131,40 @@ tooltip (visible when the mouse moves over the cell) explains the error. A
+To avoid entering exactly the same value several times in the table, it is also
+possible to create groups of samples. Within a group, a master sample can be
+designated. The values of the parameters of the master sample will be used when
+processinga all rows in the group.
```
trailing `a`

gui-co replied:
Done

gvardany commented on docs/source/interfaces/DrILL.rst:
```diff
@@ -131,6 +131,40 @@ tooltip (visible when the mouse moves over the cell) explains the error. A
+(again, one cell is sufficient) and press Ctrl + M or use the context menu.
+Grouped samples will appear with a specific label in the table. The master
+of a group will have a bold label. One can also ungroup samples by selecting
+them and pressing Ctrl + Alt + G or using the context menu.
```
I'd add here that there can be only one master row per group.

gui-co replied:
Done.

gvardany commented on docs/source/interfaces/DrILL.rst:
```diff
@@ -131,6 +131,40 @@ tooltip (visible when the mouse moves over the cell) explains the error. A
+1(master)  g1     v1           v2
+2          g1
+3          g1                  v2'
+=========  =====  ===========  ===========
```
I think we should also be able to handle a case, when for a slave row we will need to disable a given correction. In the example above, what if for sample 2 we don't want to use the v1 as parameter1 value. Is there an easy way of doing this? Of course, one could manually write the default value of the parameter in order overwrite it. But for some corrections the default is actually empty. Maybe we should have a symbol for deliberately empty cells.

gvardany reviewed (commented):
Few minor comments on the documentation. Could you add a release note as well? I will proceed to functional tests now.

gvardany commented:
It didn't work for me with the json you provide; it didn't seem to load the groups, can you check?
But I could manage to create groups and masters and process them.
I think we need also a way to add row (s) to an existing group, perhaps via the context menu.
Also we are missing the process group (s) in the menu.
For the tooltip of process rows, could you make (s) in parentheses so that it is consistent with the rest?
We are also missing, not related to the groups, create new rundex action in menu (possibly also the toolbar), which should start a fresh table from scratch, and if it wasn't empty, it should first ask if we want to save the changes or discard.
I also observed one issue which I am not sure is intended.
Create 3 rows, group them, then ungroup the first one. Then set the second as master, this changes the row labels. Is this expected?
Other than that this is a great feature and should probably become the primary way of using the interface.

darominski commented:
The example works as described however I have found a few issues elsewhere:
1. Naming of the ungrouped row reverts back to its group name and appends '1' to it, for example if one ungroups the two row of group 'A' from the example, both old row 'A1' and 'A2' will obtain the same name: 'A1'. Bolding of the former master row is also persistent. This can lead to confusion, as if one creates a new group it will also be a group 'A', so there will be multiple rows with the name name scheme and seemingly multiple master rows. The underlying information flow is fine though, so this is only on the user interface.
2. Deleting a row that is a member of the group leads to a ValueError, and DrILL still expects these rows to exist: 
`<Interface.ui.drill.model.DrillSample.DrillSample object at 0x7fb621abdb00> is not in list`
To reproduce the issue delete row 'A2' from the provided example and right-click on any cell in the interface. The issue does not occur when non-grouped row is deleted.

Then, rather than an issue but a thing of convenience: to group rows it was sufficient to highlight one cell in a row to mark them for grouping, while to delete rows one has to highlight the entire row for the delete to accept the selection. Is it an intended feature to ensure that a user is certain about deleting the row(s)?

gui-co commented:
Concerning @gvardany comments:
* the json file is modified the PR description. This one should work properly
* I modified the tooltip (`Process selected row` -> `Process selected row(s)`)
* a new action in the `File` menu is added to create a new empty table
* to solve the last issue, I modified the way the labels are updated. Whenever a group is modified, the labels are now updated to keep a uniform numbering of rows
* a release note has been added
* one can now process the selected group(s) through the menu bar
* there is an option in the context menu to add a row in an existing group

I then solved the two problems pointed by @darominski. For the last comment about the deletion of row(s), it was an intended feature but can be changed if needed.

gvardany commented on docs/source/interfaces/DrILL.rst:
```diff
@@ -131,6 +131,42 @@ tooltip (visible when the mouse moves over the cell) explains the error. A
+of a group will have a bold label. There can be only one master row per group,
+if a second row is selected as the master row, it will replace the previous one.
+One can also ungroup samples by selecting them and pressing Ctrl + Alt + G or
+using the context menu.
```
You could add that one can add a row to an existing group.

gui-co replied:
Done

gvardany commented:
Works fine, I just got an error when deleting rows (see the screenshot):

Here is the stack trace:
```
Traceback (most recent call last):
  File "/Users/vardanyan/mantid/scripts/Interface/ui/drill/model/DrillModel.py", line 788, in deleteSample
    for group in self.groups:
RuntimeError: dictionary changed size during iteration
```

![Screenshot 2021-01-12 at 08 47 35](https://user-images.githubusercontent.com/16865759/104285806-6f1dc300-54b4-11eb-92ed-5678f41399a1.png)

What is missing is to address a point I raised earlier. When we have a master row that has one column non-empty, it will be used for all the slave rows in the same group. It should be possible for the slave rows to override the parameter to be empty; that is, use the default. Perhaps the best way is to have a reserved word, e.g. DEFAULT, in which case it should use the default no matter the master row, neither the global setting.

gui-co commented:
I corrected the bug when removing the last row of a group.

I also added this special DEFAULT value for parameters in the table.

gvardany reviewed (approved):
Looks good.
